Frameworks/UIKit/UILabel.mm

```diff
@@ -44,61 +44,70 @@ @implementation UILabel {
     UILineBreakMode _lineBreakMode;
     BOOL _adjustFontSize;
     float _minimumFontSize;
-    float _originalFontSize;
     int _numberOfLines;
     BOOL _isDisabled;
     BOOL _isHighlighted;
     UIBaselineAdjustment _baselineAdjustment;
     StrongId<WXCTextBlock> _textBlock;
 }
 
-- (void)adjustFontSizeToFit {
+- (void)_adjustFontSizeToFit {
     if (_numberOfLines == 0) {
-        [self adjustTextLayerSize];
+        [self _adjustTextLayerSize];
         return;
     }
 
-    float curFontSize = _originalFontSize;
-    _font = [_font fontWithSize:curFontSize];
-
-    CGRect rect;
-    rect = [self bounds];
-    if (rect.size.width == 0.0f || rect.size.height == 0.0f) {
+    CGRect rect = [self bounds];
+    if (rect.size.width == 0.0f || rect.size.height == 0.0f || _text == nil) {
         return;
     }
 
-    while (curFontSize > _minimumFontSize && curFontSize > 0.0f) {
-        CGSize size = CGSizeZero;
+    // MaxFitFontSize records currently found maximum font size that fits the bounds,
+    // It starts with _minimumFontSize because it is smallest font size allowed by the UILabel
+    float maxFitFontSize = _minimumFontSize;
+
+    float curFontSize = [_font pointSize];
+    // Start our trial with fonts bigger than _minimumFontSize
+    if (curFontSize < maxFitFontSize) {
+        curFontSize = 2.0 * maxFitFontSize;
+        _font = [_font fontWithSize:curFontSize];
+    }
 
-        //  A single line of text should be clipped
+    // Do binary search on maximum font size that fits the bound
+    do {
+        // Caculating rect size using font and line break settings
+        CGSize size = CGSizeZero;
         if (_numberOfLines == 1) {
-            size = [_text sizeWithFont:_font constrainedToSize:CGSizeMake(rect.size.width, 0.0f) lineBreakMode:UILineBreakModeClip];
+            size = [_text sizeWithFont:_font constrainedToSize:CGSizeMake(0.0f, 0.0f) lineBreakMode:UILineBreakModeClip];
```

Comment:
because this function intend to fit the string to rect without truncation, even if you set linbreakMode as trailing or any other truncations, we still should not truncate the string with larger potentially font, but should try to fit the string with smaller font without truncation if possible.

---
In reply to: [97389805](https://github.com/Microsoft/WinObjC/pull/1768#discussion_r97389805) [](ancestors = 97389805)